File: Data Structures & Algorithms/house-robber-ii/submission-3.cpp
class Solution {
   
    vector<int>memo2;

public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];

        vector<int> memo(nums.size(),-1);
        vector<int> memo2(nums.size(),-1);
        return max(dfs(0, nums.size()-1,nums,memo), dfs(1, nums.size(),nums,memo2));
    }

private:
    int dfs(int i, int n,vector<int>& nums,vector<int> &memo) {
       if(i>=n)return 0;
       if(memo[i]!=-1)return memo[i];
       memo[i] = max(nums[i]+dfs(i+2,n,nums,memo),dfs(i+1,n,nums,memo));
       return memo[i];

    }     
};

File: Data Structures & Algorithms/interleaving-string/submission-3.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int n = s1.length();
        int m = s2.length();
        int k = s3.length();
        Boolean[][] dp = new Boolean[m+1][n+1];
        if(n+m!=k){
          return false;
        }
        return dfs(0,0,0,s1,s2,s3,dp);
    }
    boolean dfs(int i,int j,int k,String s1, String s2, String s3,Boolean[][] dp){
        if(k==s3.length()){
            return (i==s1.length() ) && (j==s2.length());
        }
        if(dp[i][j]!=null)return dp[i][j];
        boolean res = false;
        if(i<s1.length() && s1.charAt(i)==s3.charAt(k)){
            res = dfs(i+1,j,k+1,s1,s2,s3,dp);
        }
        if(j<s2.length() && s2.charAt(j)==s3.charAt(k)){
            res = dfs(i,j+1,k+1,s1,s2,s3,dp);
        }
        dp[i][j] = res;
return res;
        
    }

}
